src/scene/layer-composition.js

```diff
@@ -89,155 +111,187 @@ LayerComposition.prototype._update = function () {
+                this._dirtyBlend = this._dirtyBlend || addUniqueMeshInstance(this._meshInstances, this._meshInstancesSet, layer.opaqueMeshInstances);
+                this._dirtyBlend = this._dirtyBlend || addUniqueMeshInstance(this._meshInstances, this._meshInstancesSet, layer.transparentMeshInstances);
             }
-        }
 
-        for (i = 0; i < len; i++) {
-            this.layerList[i]._dirty = false;
-            this.layerList[i]._version++;
+            layer._dirty = false;
+            layer._version++;
         }
+
         this._dirty = false;
     }
 
+    // function adds meshInstances from src array into opaque or transparent arrays based on material blend type
+    function splitByBlendType(destOpaque, destTransparent, src) {
+        var material;
+        var srcLen = src.length;
+        for (var s = 0; s < srcLen; s++) {
+            material = src[s].material;
+            if (material && material.blendType !== BLEND_NONE) {
+                destTransparent.push(src[s]);
+            } else {
+                destOpaque.push(src[s]);
+            }
+        }
+    }
+
+    // for each layer, split its meshInstances to either opaque or transparent array based on material blend type
     if (this._dirtyBlend) {
-        // TODO: make it fast
         result |= COMPUPDATED_BLEND;
-        var opaqueOld, transparentOld, opaqueNew, transparentNew;
+
         for (i = 0; i < len; i++) {
             layer = this.layerList[i];
-            if (layer.passThrough) continue;
-            opaqueOld = layer.opaqueMeshInstances;
-            transparentOld = layer.transparentMeshInstances;
-            opaqueNew = [];
-            transparentNew = [];
-            for (j = 0; j < opaqueOld.length; j++) {
-                if (opaqueOld[j].material && opaqueOld[j].material.blendType !== BLEND_NONE) {
-                    transparentNew.push(opaqueOld[j]);
-                } else {
-                    opaqueNew.push(opaqueOld[j]);
-                }
-            }
-            for (j = 0; j < transparentOld.length; j++) {
-                if (transparentOld[j].material && transparentOld[j].material.blendType !== BLEND_NONE) {
-                    transparentNew.push(transparentOld[j]);
-                } else {
-                    opaqueNew.push(transparentOld[j]);
-                }
-            }
-            layer.opaqueMeshInstances.length = opaqueNew.length;
-            for (j = 0; j < opaqueNew.length; j++) {
-                layer.opaqueMeshInstances[j] = opaqueNew[j];
-            }
-            layer.transparentMeshInstances.length = transparentNew.length;
-            for (j = 0; j < transparentNew.length; j++) {
-                layer.transparentMeshInstances[j] = transparentNew[j];
+            if (!layer.passThrough) {
+
+                // get all instances into single array
+                var layerInstances = layer.opaqueMeshInstances.concat(layer.transparentMeshInstances);
```

Comment:
hah yeah, that's what I wrote initially, to avoid the extra copy, and then spent an hour debugging why stuff does not work. I cannot assign new array to layer.opaqueMeshInstances nor layer.transparentMeshInstances, as they're part of another class that is passed around. So whatever I do, I need to update those arrays. And so I copy data out to single array first, and then refill those arrays.

Reply:
notice the previous solution does the extra copy for the same reason as well

Reply:
I see. How about shuffling the meshInstances instead? Move the opaque meshes from transparentMeshInstances -> opaqueMeshInstance and visa-versa?

Reply:
That's an option, but I think I'm ok with the current solution as well ..

Reply:
I've done this as well ... and it's great! thanks for the suggestion.